### PR TITLE
Bug: Git plugin always treats AwsCredentials as an SSHUserPrivateKey

### PR DESCRIPTION
**Pull request by chriskilding**
Fixes [JENKINS-60329](https://issues.jenkins-ci.org/browse/JENKINS-60329) (the multi-type AwsCredentials object is interpreted as the wrong type by some downstream plugins).

Fixes rendering of credential names and descriptions in the credentials list view.

Fixes population of credentials binding dropdown boxes in the job builder and pipeline snippet generator UIs, so that only credentials of the appropriate type are shown.

**WARNING:** This fix requires a new tag `jenkins:credentials:type` (see README for instructions) to be present on Secrets Manager secrets that you want to use from Jenkins. This is a breaking change: if the tag is not present, Jenkins will not make that secret available as a credential.

### CODE REVIEW

**Comment by chriskilding** [comment]
@mhaley-miovision If you want to test this, the functionality is all there and it does work on my laptop - the only build breaker is a `java.net.SocketTimeoutException: Read timed out` error in the Jenkins core test suite, which only happens on Java 8.

File: src/test/java/io/jenkins/plugins/credentials/secretsmanager/util/git/GitHttpServer.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                server.stop();
+            } catch (Exception e) {
+                // Don't care
+            }
+        }
+    }
+
+    public String getCloneUrl(String repo, String username) {
+        return server.getUri().toASCIIString();
+        //return String.format("ssh://%s@localhost:%d/%s.git", username, port, repo);
```
Comment (mhaley-miovision) [inline src/test/java/io/jenkins/plugins/credentials/secretsmanager/util/git/GitHttpServer.java]: Remove unused comment.

File: src/test/java/io/jenkins/plugins/credentials/secretsmanager/GitPluginIT.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        private final String repo = "foo";
+        private final String username = "agitter";
+
+        @Rule
+        public final GitHttpServer git = new GitHttpServer(Collections.singletonList(repo));
+
+        @Test
+        @ConfiguredWithCode(value = "/integration.yml")
+        public void shouldSupportGitPlugin() throws Exception {
+            // FIXME this works on the master, but not on an agent - something calls Jenkins.getInstance on the agent & it's null
```
Comment (mhaley-miovision) [inline src/test/java/io/jenkins/plugins/credentials/secretsmanager/GitPluginIT.java]: I ran into a problem like this myself.  Basically the credential snapshot was not self contained and when the agent called getPassword it was not decrypted yet.

File: src/test/java/io/jenkins/plugins/credentials/secretsmanager/FiltersIT.java [hunk 3]
```diff
@@ -19,11 +18,15 @@
...
-            opts.tags = Collections.singletonMap("product", "roadrunner");
+            opts.tags = Maps.of(
+                    "jenkins:credentials:type", "string",
+                    "product", "roadrunner");
         });
         // And
         final CreateSecretOperation.Result bar = createOtherSecret("supersecret", opts -> {
-            opts.tags = Collections.singletonMap("product", "coyote");
+            opts.tags = Maps.of(
+                    "jenkins:credentials:type", "string",
```
Comment (mhaley-miovision) [inline src/test/java/io/jenkins/plugins/credentials/secretsmanager/FiltersIT.java]: Can we create a static string to hold this value and then just reference it here?  A good spot for this static string may be the `CredentialsFactory` class.

File: src/main/java/io/jenkins/plugins/credentials/secretsmanager/factory/CredentialsFactory.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        private SecretSupplier(AWSSecretsManager client, String name) {
+            super(client, name);
+        }
+
+        @Override
+        public Secret get() {
+            return getSecretValue().match(new SecretValue.Matcher<Secret>() {
+                @Override
+                public Secret string(String str) {
+                    return Secret.fromString(str);
```
Comment (mhaley-miovision) [inline src/main/java/io/jenkins/plugins/credentials/secretsmanager/factory/CredentialsFactory.java]: This was what caused my null pointer problem.  Basically the call to Secret.fromString was being delayed until the agent called getPassword on the credential snapshot and then it blows up.

File: src/main/java/io/jenkins/plugins/credentials/secretsmanager/factory/username_password/AwsUsernamePasswordCredentials.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    public AwsUsernamePasswordCredentials(String id, String description, Supplier<Secret> password, String username) {
+        super(id, description);
+        this.password = password;
+        this.username = username;
+    }
+
+    @NonNull
+    @Override
+    public Secret getPassword() {
+        return password.get();
```
Comment (mhaley-miovision) [inline src/main/java/io/jenkins/plugins/credentials/secretsmanager/factory/username_password/AwsUsernamePasswordCredentials.java]: Your snapshot of this should call `password.get()` in the constructor I think.

File: src/test/java/io/jenkins/plugins/credentials/secretsmanager/GitPluginIT.java (hunk 2 above)
Comment (chriskilding) [reply src/test/java/io/jenkins/plugins/credentials/secretsmanager/GitPluginIT.java]: I'd forgotten to remove that comment

File: src/test/java/io/jenkins/plugins/credentials/secretsmanager/FiltersIT.java (hunk 3 above)
Comment (chriskilding) [reply src/test/java/io/jenkins/plugins/credentials/secretsmanager/FiltersIT.java]: Good idea, I've formalised two classes to contain the constants - `Tags` and `Type`.

File: src/main/java/io/jenkins/plugins/credentials/secretsmanager/factory/CredentialsFactory.java (hunk 4 above)
Comment (chriskilding) [reply src/main/java/io/jenkins/plugins/credentials/secretsmanager/factory/CredentialsFactory.java]: Yep, that one was hard to find. Did you mean that my changes will fix it, or that they currently still cause an NPE?

File: src/main/java/io/jenkins/plugins/credentials/secretsmanager/factory/username_password/AwsUsernamePasswordCredentials.java (hunk 5 above)
Comment (chriskilding) [reply src/main/java/io/jenkins/plugins/credentials/secretsmanager/factory/username_password/AwsUsernamePasswordCredentials.java]: The idea is the CredentialSnapshotTaker classes are responsible for resolving the secret value. (CredentialsSnapshotTaker is the idiomatic place for this to happen in the Jenkins credentials API.) They then construct a new version of the credential with a constant-value supplie that just returns the value (no calls to AWS). This is then handed off to the agent.

**Comment by chriskilding** [comment]
It looks like there is a flaky test in `org.jvnet.hudson.test.JellyTestSuiteBuilder$JellyTestSuite` - sometimes we get a socket read timeout, sometimes not.

I'm not clear on whether the plugin's own code is the cause or something in the JellyTestSuite. (In this PR I have bumped the parent POM version in a bid to resolve dependency enforcer problems, so maybe a newer version of Jenkins test harness has a bug.)

**Comment by mhaley-miovision** [comment]
Ok, I was able to build this locally to get the .hpi file.  I'll be standing up a jenkins instance today for testing out our deployment process, and can get this plugin installed too and give it a test.  I will let you know if I hit any problems.

Thanks for the quick turn around here.

File: docs/README.md [hunk 6]
```diff
@@ -86,13 +96,24 @@ node {
...
+- `jenkins:credentials:type` = `usernamePassword`
+- `jenkins:credentials:username` = *username*
+
+#### Example
+
+AWS CLI:
 
 ```bash
-aws secretsmanager create-secret --name 'artifactory' --secret-string 'supersecret' --tags 'Key=jenkins:credentials:username,Value=joe' --description 'Acme Corp Artifactory login'
+aws secretsmanager create-secret --name 'artifactory' --secret-string 'supersecret' --tags 'Key=jenkins:credentials:type,Value=usernamePassword,jenkins:credentials:username,Value=joe' --description 'Acme Corp Artifactory login'
```
Comment (mhaley-miovision) [inline docs/README.md]:
```suggestion
aws secretsmanager create-secret --name 'artifactory' --secret-string 'supersecret' --tags 'Key=jenkins:credentials:type,Value=usernamePassword,Key=jenkins:credentials:username,Value=joe' --description 'Acme Corp Artifactory login'
```
Just caught this today.

File: docs/README.md [hunk 7]
```diff
@@ -86,13 +96,24 @@ node {
...
+- `jenkins:credentials:type` = `usernamePassword`
+- `jenkins:credentials:username` = *username*
+
+#### Example
+
+AWS CLI:
 
 ```bash
-aws secretsmanager create-secret --name 'artifactory' --secret-string 'supersecret' --tags 'Key=jenkins:credentials:username,Value=joe' --description 'Acme Corp Artifactory login'
+aws secretsmanager create-secret --name 'artifactory' --secret-string 'supersecret' --tags 'Key=jenkins:credentials:type,Value=usernamePassword,Key=jenkins:credentials:username,Value=joe' --description 'Acme Corp Artifactory login'
```
Comment (mhaley-miovision) [inline docs/README.md]: So testing out a little more and I found that you need to space separate each of the tags for all of them to apply.  I would give this a quick test on your machine to make sure it does what you expect.
```suggestion
aws secretsmanager create-secret --name 'artifactory' --secret-string 'supersecret' --tags 'Key=jenkins:credentials:type,Value=usernamePassword' 'Key=jenkins:credentials:username,Value=joe' --description 'Acme Corp Artifactory login'
```

File: docs/README.md [hunk 8]
```diff
@@ -120,23 +141,25 @@ node {
...
-- **Encoding**
-  - PKCS#1 (starts with `-----BEGIN [ALGORITHM] PRIVATE KEY-----`)
-  - PKCS#8 (starts with `-----BEGIN PRIVATE KEY-----`)
-  - OpenSSH (starts with `-----BEGIN OPENSSH PRIVATE KEY-----`)
+AWS CLI:
 
 ```bash
 ssh-keygen -t rsa -b 4096 -C 'acme@example.com' -f id_rsa
-aws secretsmanager create-secret --name 'ssh-key' --secret-string 'file://id_rsa' --tags 'Key=jenkins:credentials:username,Value=joe' --description 'Acme Corp SSH key'
+aws secretsmanager create-secret --name 'ssh-key' --secret-string 'file://id_rsa' --tags 'Key=jenkins:credentials:type,Value=sshUserPrivateKey,Key=jenkins:credentials:username,Value=joe' --description 'Acme Corp SSH key'
```
Comment (mhaley-miovision) [inline docs/README.md]: Same as above, give this a quick test before committing.
```suggestion
aws secretsmanager create-secret --name 'ssh-key' --secret-string 'file://id_rsa' --tags 'Key=jenkins:credentials:type,Value=sshUserPrivateKey' 'Key=jenkins:credentials:username,Value=joe' --description 'Acme Corp SSH key'
```

**Comment by chriskilding** [comment]
I eliminated the flaky test by rolling back the parent POM version (and using a much older JGit HTTP test dependency version). 2 passing builds in succession suggests it's ready for a wider review.

**Comment by chriskilding** [comment]
CC @jenkinsci/code-reviewers

File: src/main/java/io/jenkins/plugins/credentials/secretsmanager/factory/CredentialsFactory.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+public abstract class CredentialsFactory {
+
+    private CredentialsFactory() {
+
+    }
+
+    /**
+     * Construct a Jenkins credential from a Secrets Manager secret.
+     *
+     * @param name the secret's name (must be unique)
```
Comment (jeffpearce) [inline src/main/java/io/jenkins/plugins/credentials/secretsmanager/factory/CredentialsFactory.java]: Nit: might be useful to be specific - I believe it needs to be unique within an AWS account?

Comment (chriskilding) [reply src/main/java/io/jenkins/plugins/credentials/secretsmanager/factory/CredentialsFactory.java]: Updated the doc

**Comment by mhaley-miovision** [comment]
Awesome, thanks for all the work here @chriskilding.

**Comment by c4m4** [comment]
I am facing this problem, which git plugin version do I need to use to solve this bug?

**Comment by MarkEWaite** [comment]
> I am facing this problem, which git plugin version do I need to use to solve this bug?

No change was made to the git plugin for this issue.  The current version should work as well as any other version.  The AWS secrets manager credentials provider plugin was modified by this pull request to report the AWS credentials as an SSH user private key.

**Pull request merged**